### ISSUE TITLE
Invalid "formula" key causes silent crash

### ISSUE THREAD

**Issue by Dalas121** [wesnoth/wesnoth#8522]
### Game and System Information

- **Version:** 1.17.26
- **Downloaded from:** Steam
- **OS:** Windows

### Description of the bug

Targeting a weapon attack with the following invalid specials causes Wesnoth to crash without an error.

The formula has an extra closing parenthesis, which causes the crash. Crashing is fine, error-less is the bug.

```
            [damage_type]
                [filter_opponent]
                    formula="$( $self[arcane] >= max([ $self[blade], $self[impact], $self[pierce], $self[arcane], $self[cold], $self[fire] ]) ))"
                [/filter_opponent]
                alternative_type=arcane
            [/damage_type]
            [damage_type]
                [filter_opponent]
                    formula="$( $self[fire] >= max([ $self[blade], $self[impact], $self[pierce], $self[arcane], $self[cold], $self[fire] ]) ))"
                [/filter_opponent]
                alternative_type=fire
            [/damage_type]
            [damage_type]
                [filter_opponent]
                    formula="$( $self[cold] >= max([ $self[blade], $self[impact], $self[pierce], $self[arcane], $self[cold], $self[fire] ]) ))"
                [/filter_opponent]
                alternative_type=cold
            [/damage_type]
```

### Steps to reproduce the behavior

1. Assign the listed specials to a unit's weapon
2. Try to make an attack with the unit

### Expected behavior

Wesnoth should throw an error, recognizing the invalid `formula` syntax.

**Comment by cooljeanius** [comment]
semi-related: #8385

**Comment by newfrenchy83** [comment]
The alternative_type choose automaticelly the damage type who done more damage but for the weapon applied, here i see what you use formula self[cold] >max[cold] perhaps what you should remove max damage when same type what damage compared

**Comment by newfrenchy83** [comment]
Oh pardon, i dont undertstand what bug is intentional but not behavior of game when error in formula, have you trier with other weapn special ?

**Comment by newfrenchy83** [comment]
Like alternative_type select the damage type who done more damage between original and [damage_type]alternative_type= why you encode this in [filter_oppoonent]?

**Comment by newfrenchy83** [comment]
I have tried several times and the problem is not specific to [damage_type], I don't know why wesnoth crashes instead of displaying a syntax error message and I don't see how to fix it

**Comment by CelticMinstrel** [comment]
> Crashing is fine

No. No it's not. Crashing is _never_ fine, not even if you've coded something _incredibly dumb_ (which yes, this is). There should _always_ be an error message instead of a crash.

> ```
> [damage_type]
>     [filter_opponent]
>         formula="$( $self[arcane] >= max([ $self[blade], $self[impact], $self[pierce], $self[arcane], $self[cold], $self[fire] ]) ))"
>     [/filter_opponent]
>     alternative_type=arcane
> [/damage_type]
> [damage_type]
>     [filter_opponent]
>         formula="$( $self[fire] >= max([ $self[blade], $self[impact], $self[pierce], $self[arcane], $self[cold], $self[fire] ]) ))"
>     [/filter_opponent]
>     alternative_type=fire
> [/damage_type]
> [damage_type]
>     [filter_opponent]
>         formula="$( $self[cold] >= max([ $self[blade], $self[impact], $self[pierce], $self[arcane], $self[cold], $self[fire] ]) ))"
>     [/filter_opponent]
>     alternative_type=cold
> [/damage_type]
> ```

Uhhhhhhh… what on earth are you **_doing_** here? The extra closing parenthesis is just the _start_ of the errors in that snippet.

1. Extra closing parenthesis bad, obviously
2. `$self[arcane]` is not a valid WML variable substitution (though the variable parser is rather forgiving, so I guess there's a chance that it parses it equivalent to `self.arcane` – something which in my opinion shouldn't be supported, mainly just to keep things simple)
3. You're using a formula substitution (`$()`) in `formula=` - there is no reason to ever do this

The correct formula would be something like this (assuming no custom damage types are in play):

```
formula="resistance.fire >= max(values(resistance))"
```

Though it can never be greater, so you might as well delete the `>` as well. If custom damage types _are_ in play and you only want to check the default ones, you can write out all the `resistance.whatever` inside the `max()` in place of `values(resistance)`.

Though obviously all of that is assuming that you want "self" to refer to the opponent and not some random other unit on the map. Still, if you _do_ want "self" to refer to some random other unit on the map, and and `$self` is a stored unit, I think all you need to do is append `where self = unit_at($self.x, $self.y)` to the formula.

**Comment by Dalas121** [comment]
Yeah, the snippet is very wrong. Right now I've got it all working correctly, I appreciate the notes.
```
#define ALTERNATIVE_DAMAGE_TYPE TYPE
    [damage_type]
        [filter_opponent]
            formula=" self.resistance['{TYPE}'] <= min([ self.resistance['blade'], self.resistance['impact'], self.resistance['pierce'], self.resistance['arcane'], self.resistance['cold'], self.resistance['fire'] ]) "
        [/filter_opponent]
        alternative_type={TYPE}
    [/damage_type]
#enddef
```

**Comment by CelticMinstrel** [comment]
I think `self.resistance.blade` should also work – from what I remember, WFL allows you to address maps "as if" they were an object as long as the keys are strings that are valid identifiers.

And as mentioned in the previous comment, the `values(self.resistance)` thing would still simplify your formula if you don't mind it also catching custom damage types.

**Comment by Dalas121** [comment]
Will do, thanks.  Much simpler for sure, and I don't mind catching custom damage types.

**Comment by newfrenchy83** [comment]
> Yeah, the snippet is very wrong. Right now I've got it all working correctly, I appreciate the notes.
> 
> ```
> #define ALTERNATIVE_DAMAGE_TYPE TYPE
>     [damage_type]
>         [filter_opponent]
>             formula=" self.resistance['{TYPE}'] = min([ self.resistance['blade'], self.resistance['impact'], self.resistance['pierce'], self.resistance['arcane'], self.resistance['cold'], self.resistance['fire'] ]) "
>         [/filter_opponent]
>         alternative_type={TYPE}
>     [/damage_type]
> #enddef
> ```
> 
> ```
> [dummy]
>     name=_"astral"
>     description=_"This attack always deals whichever damage type its opponent is most vulnerable to."
> [/dummy]
> {ALTERNATIVE_DAMAGE_TYPE cold}
> {ALTERNATIVE_DAMAGE_TYPE fire}
> {ALTERNATIVE_DAMAGE_TYPE arcane}
> {ALTERNATIVE_DAMAGE_TYPE blade}
> {ALTERNATIVE_DAMAGE_TYPE pierce}
> {ALTERNATIVE_DAMAGE_TYPE impact}
> ```

I see that you use the alternative_type attribute in [damage_type] but do you know that if you use it instead of replacement_type then you do not need a formula because the resistance comparison is programmed in the source code and alternative_type chosen always the type for which the opponent is least resistant

**Comment by newfrenchy83** [comment]
if you need to choice the alternative_type in list of many damage_type, then yes use formula for exclude of list used the damage_type who don't matches is a good method, the altenative_type used alone is valable for a only [damage_type] else this isthe [damage_type] in head of list who will be used.
